### ISSUE TITLE
[java] [regression from 6.55] LooseCoupling false-positive

### ISSUE THREAD

**Issue by kohlschuetter** [pmd/pmd#4615]
**Affects PMD Version:** 
7.0.0-SNAPSHOT

**Rule:** [LooseCouling](https://docs.pmd-code.org/latest/pmd_rules_java_bestpractices.html#loosecoupling)

**Description:**

**Code Sample demonstrating the issue:**

```java
private static final WeakHashMap<ByteBuffer, @Nullable AFTIPCTopologySubscription> SUBSCRIPTIONS =
      new WeakHashMap<>();
```
or
```java
final Map<String, LinkedHashSet<Class<?>>> testMap = new LinkedHashMap<>();
```

**Expected outcome:**

PMD reports a violation, but that's wrong (regression). That's probably a false positive:

`WeakHashMap` (just like `LinkedHashMap` / `LinkedHashSet`) carries additional information over a simple `HashMap`. These implementations change the expected behavior of `HashMap` (elements can disappear, or, with LinkedHashMap are stored in a certain order). Since this wasn't reported before in 6.55, I'd prefer the old behavior.

**Running PMD through:** CLI/Maven

**Comment by adangel** [comment]
I had a look at what happens here. With PMD 6.55.0, we only considered a explicit list of collection classes: https://github.com/pmd/pmd/blob/ef3455348603aa25f86894b9930f05f141f44d20/pmd-core/src/main/java/net/sourceforge/pmd/util/CollectionUtil.java#L38-L41

And WeakHashMap was not in the list. With PMD 7, we now consider all collection types, that means all types that inherit/implement java.util.Collection or java.util.Map. Hence we now detect this case.

Overall, I don't think, this is a false positive. Whether you use WeakHashMap or another Map implementation is IMHO an implementation detail. And the map should not be exposed, as you pointed out, that this map may behave strangely. You already declared the map as "private", so as long as you don't write a getter for that, you should be fine.
Also note, from the javadoc of WeakHashMap https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/WeakHashMap.html:

> This class is intended primarily for use with key objects whose equals methods test for object identity using the == operator. Once such a key is discarded it can never be recreated, so it is impossible to do a lookup of that key in a WeakHashMap at some later time and be surprised that its entry has been removed. This class will work perfectly well with key objects whose equals methods are not based upon object identity, such as String instances. With such recreatable key objects, however, the automatic removal of WeakHashMap entries whose keys have been discarded may prove to be confusing. 

You key type is "ByteBuffer" which doesn't use object identity for equals. That's why you might experience that values are disappearing.

As long as there are no extra methods that are not on java.util.Map, there is no problem I see. For WeakHashMap, that is the case. It appears as a Map. Other we would run into #1218.

If you consider java.util.Collections#synchronizedMap - this also returns just a Map and not a special type, that transports the information, that the map is synchronized.

If you want to exclude WeakHashMap for you project, you need to configure the rule:

```xml
<rule ref="category/java/bestpractices.xml/LooseCoupling">
    <properties>
        <property name="allowedTypes" value="java.util.Properties,java.util.WeakHashMap" />
    </properties>
</rule>
```

-----

Now, the next problem: `final Map<String, LinkedHashSet<Class<?>>> testMap = new LinkedHashMap<>();`

In PMD 6.55.0 we didn't consider the generic types and so missed out LinkedHashSet. With PMD 7, we also find this. Again, I don't think, this is a false positive - this is what this rule is supposed to do: Don't use specific implementations.

You can again configured "allowedTypes" to exclude LinkedHashSet as well.

But with Java 21, there is another possibility: The new [JEP 431: Sequenced Collections](https://openjdk.org/jeps/431) introduced the new type [java.util.SequencedCollection](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/SequencedCollection.html) which transports the information, that the elements in the collection have a specific order. You can use that:

```java
final Map<String, SequencedCollection<Class<?>>> testMap = new LinkedHashMap<>();
```

-----

In summary, I think, both cases are not false positive but exactly the cases, the rule should detect. That these cases were not found in PMD 6.55.0 are actually false negatives, which have now been fixed with PMD 7.

I'm closing this as not planned. Feel free to comment/reopen.
